opa/stackable/bin/process-logs

```diff
@@ -0,0 +1,145 @@
+            ;;
+        --opa-rolling-log-files)
+            shift
+            OPA_ROLLING_LOG_FILES=$1
+            ;;
+        --stackable-log-dir)
+            shift
+            STACKABLE_LOG_DIR=$1
+            ;;
+        --container-name)
+            shift
+            CONTAINER_NAME=$1
+            ;;
+        *)
+            break
+            ;;
+    esac
+    shift
+  done
+
+  if [ -z "${CONSOLE_LEVEL}" ]; then
+    echo "ERROR: parameter --console-log-level cannot be empty"
+    exit 1
+  fi
+
+  if [ -z "${FILE_LEVEL}" ]; then
+    echo "ERROR: parameter --file-log-level cannot be empty"
+    exit 1
+  fi
+
+  if [ -z "${DECISION_LEVEL}" ]; then
+    echo "ERROR: parameter --decision-log-level cannot be empty"
+    exit 1
+  fi
+
+  if [ -z "${SERVER_LEVEL}" ]; then
+    echo "ERROR: parameter --server-log-level cannot be empty"
+    exit 1
+  fi
+
+  if [ -z "${OPA_ROLLING_LOG_FILE_SIZE_BYTES}" ]; then
+    echo "ERROR: parameter --opa-rolling-log-file-size-bytes cannot be empty"
+    exit 1
+  fi
+
+  if [ -z "${OPA_ROLLING_LOG_FILES}" ]; then
+    echo "ERROR: parameter --opa-rolling-log-files cannot be empty"
+    exit 1
+  fi
+
+  if [ -z "${STACKABLE_LOG_DIR}" ]; then
+    echo "ERROR: parameter --stackable-log-dir cannot be empty"
+    exit 1
+  fi
+
+  if [ -z "${CONTAINER_NAME}" ]; then
+    echo "ERROR: parameter --container-name cannot be empty"
+    exit 1
+  fi
+}
+
+get_levels() {
+    case $1 in
+        fatal)
+            echo '["fatal"]' ;;
+        error)
+            echo '["error","fatal"]' ;;
+        warn)
+            echo '["warn","error","fatal"]' ;;
+        info)
+            echo '["info","warn","error","fatal"]' ;;
+        debug)
+            echo '["debug","info","warn","error","fatal"]' ;;
+        trace)
+            echo '["trace","debug","info","warn","error","fatal"]' ;;
+        *)
+            echo '[""]' ;;
+    esac
+}
+
+main() {
+    parse_args "$@"
+
+    local DECISION_LEVELS
+    DECISION_LEVELS=$(get_levels "$DECISION_LEVEL")
+    local SERVER_LEVELS
+    SERVER_LEVELS=$(get_levels "$SERVER_LEVEL")
+    local CONSOLE_LEVELS
+    CONSOLE_LEVELS=$(get_levels "$CONSOLE_LEVEL")
+    local FILE_LEVELS
+    FILE_LEVELS=$(get_levels "$FILE_LEVEL")
+
+    jq -c --unbuffered 'if .decision_id then .logger = "decision" else .logger = "server" end' |
+    jq -c --unbuffered --arg decision_levels "$DECISION_LEVELS" --arg server_levels "$SERVER_LEVELS" \
+        'select(((.logger == "decision") and (.level | inside($decision_levels))) or 
+        ((.logger == "server") and (.level | inside($server_levels))))' |
+    tee >(jq -c --unbuffered --arg file_levels "$FILE_LEVELS" 'select(.level | inside($file_levels))' \
+        > >(/stackable/multilog s"$OPA_ROLLING_LOG_FILE_SIZE_BYTES" n"$OPA_ROLLING_LOG_FILES" "$STACKABLE_LOG_DIR"/"$CONTAINER_NAME")) |
+    jq -c --unbuffered --arg console_levels "$CONSOLE_LEVELS" 'select(.level | inside($console_levels))'
```

Comment:
The pipe terminates if an error occurs. This is the case e.g. if the input is not valid JSON or the JSON object does not contain the level property. It is important that this script is fail-safe, otherwise the command calling it, will probably also fail.

This comment contradicts the proposal above that `set -euo pipefail` should be set.

Reply:
I added something that maybe combines the two things, please take a look at it and maybe @NickLarsenNZ too.

So now `-eo pipefail` is still in there for any failures other than the processing of log messages. And jq processing is wrapped in a try-catch block to catch log messages which are missing the level property and also are invalid json (by using -R and fromjson parsing). Would that solve both concerns?

Also for some reason after applying shellcheck linting suggestions to change `main $@` and `parse_args $@` to `main "$@"` and `parse_args "$@"` the -u flag throws an unbound variable error
`./opa/stackable/bin/process-logs: line 24: $1: unbound variable` which shouldn't be true? When removing the flag and running the same command again, it works fine. Maybe someone has an idea, I removed the flag for now since the unset variables are handled anyway as mentioned earlier.

Reply:
Just in addition, you can enable/disable the features around certain blocks of code.

```sh
set -euo pipefail

# code here

set +e
# failures ignored here
set -e

# failures not ignored anymore
```

Reply:
Works perfectly. Thank you!